Source/Particles/Resampling/VelocityCoincidenceThinning.cpp

```diff
@@ -0,0 +1,284 @@
+                cluster_uy += w[part_idx]*uy[part_idx];
+                cluster_uz += w[part_idx]*uz[part_idx];
+                total_weight += w[part_idx];
+                total_energy += w[part_idx] * Algorithms::KineticEnergy(
+                    ux[part_idx], uy[part_idx], uz[part_idx], mass
+                );
+
+                // check if this is the last particle in the current momentum bin
+                if (
+                    (momentum_bin_number_data[sorted_indices_data[i]] != momentum_bin_number_data[sorted_indices_data[i + 1]])
+                    || (i == cell_stop - 2) )
+                {
+                    // check if the bin has more than 2 particles in it
+                    if ( particles_in_bin > 2 && total_weight > std::numeric_limits<amrex::ParticleReal>::min() ){
+                        // get average quantities for the bin
+#if !defined(WARPX_DIM_1D_Z)
+                        cluster_x /= total_weight;
+#endif
+#if defined(WARPX_DIM_3D)
+                        cluster_y /= total_weight;
+#endif
+                        cluster_z /= total_weight;
+                        cluster_ux /= total_weight;
+                        cluster_uy /= total_weight;
+                        cluster_uz /= total_weight;
+
+                        // perform merging of momentum bin particles
+                        auto u_perp2 = cluster_ux*cluster_ux + cluster_uy*cluster_uy;
+                        auto u_perp = std::sqrt(u_perp2);
+                        auto cluster_u_mag2 = u_perp2 + cluster_uz*cluster_uz;
+                        auto cluster_u_mag = std::sqrt(cluster_u_mag2);
+
+                        // calculate required velocity magnitude to achieve
+                        // energy conservation
+                        auto v_mag2 = total_energy / total_weight * (
+                            (total_energy / total_weight + 2._prt * mass * c2 )
+                            / (mass * mass * c2)
+                        );
+                        auto v_perp = (v_mag2 > cluster_u_mag2) ? std::sqrt(v_mag2 - cluster_u_mag2) : 0_prt;
+
+                        // choose random angle for new velocity vector
+                        auto phi = amrex::Random(engine) * MathConst::pi;
+
+                        // set new velocity components based on chosen phi
+                        auto vx = v_perp * std::cos(phi);
+                        auto vy = v_perp * std::sin(phi);
+
+                        // calculate rotation angles to parallel coord. frame
+                        auto cos_theta = cluster_uz / cluster_u_mag;
+                        auto sin_theta = u_perp / cluster_u_mag;
+                        auto cos_phi = cluster_ux / u_perp;
+                        auto sin_phi = cluster_uy / u_perp;
+
+                        // handle edge case with stationary clusters or particles
+                        if (std::abs(u_perp) <= std::numeric_limits<amrex::ParticleReal>::min()) {
```

Comment:
Is there a reason to use the numeric limit min rather than the more simple `std::abs(u_perp) == 0.`?

Reply:
Just to make the protection fully capture the numerical resolution. Basically, checking whether `u_perp` is closer to zero than the numerically smallest value of `ParticleReal`.

Reply:
This is Ok, but it looks like `u_perp` will be of type `ParticleReal`, so checking for `== 0.` should be sufficient.

Reply:
I avoided the issue of comparing floats for equality by changing the logic to check if `u_perp > 0`.